geo/src/algorithm/voronoi_diagram.rs

```diff
@@ -0,0 +1,970 @@
+
+impl<T: GeoFloat> VoronoiDiagram<T> for Polygon<T>
+where
+    f64: From<T>,
+{
+    fn compute_voronoi_components(
+        &self,
+        clipping_mask: Option<&Polygon<T>>,
+    ) -> Result<VoronoiComponents<T>> {
+        compute_voronoi_components(self, clipping_mask)
+    }
+
+    fn compute_voronoi_diagram(&self, _clipping_mask: Option<&Polygon<T>>) -> Result<Vec<Polygon>> {
+        todo!("Need to map the components to voronoi cells");
+    }
+}
+impl<T: GeoFloat> VoronoiDiagram<T> for MultiPoint<T>
+where
+    f64: From<T>,
+{
+    fn compute_voronoi_components(
+        &self,
+        clipping_mask: Option<&Polygon<T>>,
+    ) -> Result<VoronoiComponents<T>> {
+        compute_voronoi_components(self, clipping_mask)
+    }
+
+    fn compute_voronoi_diagram(&self, _clipping_mask: Option<&Polygon<T>>) -> Result<Vec<Polygon>> {
+        todo!("Need to map the components to voronoi cells");
+    }
+}
+
+fn compute_voronoi_components<T: GeoFloat, U: TriangulateDelaunay<T>>(
+    polygon: &U,
+    clipping_mask: Option<&Polygon<T>>,
+) -> Result<VoronoiComponents<T>>
+where
+    f64: From<T>,
+{
+    let triangles = polygon
+        .delaunay_triangulation()
+        .map_err(VoronoiDiagramError::DelaunayError)?;
+
+    compute_voronoi_components_from_delaunay(&triangles, clipping_mask)
+}
+
+/// Compute the Voronoi Diagram from Delaunay Triangles
+/// The Voronoi Diagram is a [dual graph](https://en.wikipedia.org/wiki/Dual_graph)
+/// of the [Delaunay Triangulation](https://en.wikipedia.org/wiki/Delaunay_triangulation)
+/// and thus the Voronoi Diagram can be created from the Delaunay Triangulation.
+pub fn compute_voronoi_components_from_delaunay<T: GeoFloat>(
+    triangles: &[Triangle<T>],
+    clipping_mask: Option<&Polygon<T>>,
+) -> Result<VoronoiComponents<T>>
+where
+    f64: From<T>,
+{
+    let delaunay_triangles: Vec<DelaunayTriangle<T>> =
+        triangles.iter().map(|x| (*x).into()).collect();
+
+    if delaunay_triangles.is_empty() {
+        return Ok(VoronoiComponents {
+            delaunay_triangles: vec![],
+            vertices: vec![],
+            lines: vec![],
+            neighbours: vec![],
+        });
+    }
+
+    let clipping_mask = match clipping_mask {
+        Some(mask) => mask.clone(),
+        None => create_clipping_mask(triangles)?,
+    };
+
+    // The centres of the delaunay circumcircles form the vertices of the
+    // voronoi diagram
+    let mut vertices: Vec<Coord<T>> = Vec::new();
+    for tri in delaunay_triangles.iter() {
+        vertices.push(
+            tri.get_circumcircle()
+                .map_err(VoronoiDiagramError::DelaunayError)?
+                .center,
+        );
+    }
+
+    // Find the shared edges
+    let mut shared = find_shared_edges(&delaunay_triangles);
+
+    // Create the lines joining the vertices
+    let mut voronoi_lines: Vec<Line<T>> = Vec::new();
+    for neighbour in shared.neighbours.iter() {
+        if neighbour.0.is_none() || neighbour.1.is_none() {
```

Comment:
The unwraps can be removed using
```rust
if let (Some(x), Some(y)) = neighbour { ... }
```

Reply:
oh so much better! Don't know why that didnt occur to me, thanks @jayvdb